### ISSUE TITLE
[4.1-beta] select_nested inputs don't show up in MonsterCRUD

### ISSUE THREAD

**Issue by tabacitu** [Laravel-Backpack/CRUD#2767]
# Bug report

### What I did:

- merged the relationship changes in Backpack; I mean https://github.com/Laravel-Backpack/CRUD/pull/2632 ;
- merged Laravel-Backpack/demo#164 which adds relationship to Dummy CRUD;

### What I expected to happen:

- the exact same relationship fields to be shown both inside MonsterCRUD and DummyCRUD;

### What happened:

- some fields that were showing in MonsterCRUD before merging https://github.com/Laravel-Backpack/CRUD/pull/2632 no longer show in MonsterCRUD, but do show in DummyCRUD:

In MonsterCRUD:

![Screenshot 2020-05-05 at 11 43 50](https://user-images.githubusercontent.com/1032474/81048835-baafb400-8ec5-11ea-8839-a92684069a2d.png)

In DummyCRUD:

![Screenshot 2020-05-05 at 11 43 43](https://user-images.githubusercontent.com/1032474/81048846-bdaaa480-8ec5-11ea-8077-d975da39ddc0.png)

I suspect this is because starting with https://github.com/Laravel-Backpack/CRUD/pull/2632 relationship fields have their stuff overwritten. More exactly "name". It looks like whenever there's an ```entity``` things go haywire - Backpack overwrites stuff that I've explicitly defined.

### What I've already tried to fix it:

Nothing yet, will look at it in 30 min after lunch. But this highlights a big problem in how people might ALREADY be using Backpack. This should not be fixed in Demo, but in CRUD. 

I don't know exactly how right now. But by default we should not overwrite people's stuff. Only in very very niche use cases. If the `relationship` field is the only one that benefits from this, maybe we'll actually hardcode something like - if ```$field['type'] == 'relationship'``` then ```$field['auto'] = true``` or something like that. A flag that would let the system know whether it's ok to overwrite what the developer defined. Or maybe `$field['overwrite']` would be a better name for it.

### Backpack, Laravel, PHP, DB version:
```
### PHP VERSION:
PHP 7.4.2 (cli) (built: Jan 22 2020 06:30:58) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.2, Copyright (c), by Zend Technologies

### LARAVEL VERSION:
v7.9.2@757b155658ae6da429065ba8f22242fe599824f7

### BACKPACK VERSION:
4.1.x-dev@00150a6e5c7e527d2f3121730dcf8ed4dcc2b6b9
```

**Comment by pxpm** [comment]
I haven't gone through this, but as my eyes are vicious in this relationship work I would rather wait for you to have a look, and I am here to help you with what you want. 

My first guess is that we introduced fields with same "name" (or after relationship inner works it gets the same name), so they edit the same column in database and the first defined does not appear.

**Comment by pxpm** [comment]
Indeed, you and me were both right. 

It happens because the field name changes, but well, field name changes because it's the name that is used by laravel in the relationship. 

Eg: if you dd() `select_grouped` after `makeSureFieldHasNecessaryAttributes()` this is what you get:

```php
rray:15 [▼
  "label" => "Select_grouped"
  "type" => "select_grouped"
  "name" => "select2_from_ajax"
  "fake" => true
  "entity" => "article"
  "model" => "Backpack\NewsCRUD\app\Models\Article"
  "attribute" => "title"
  "group_by" => "category"
  "group_by_attribute" => "name"
  "group_by_relationship_back" => "articles"
  "tab" => "Relationships"
  "wrapperAttributes" => array:1 [▶]
  "relation_type" => "BelongsTo"
  "multiple" => false
  "pivot" => false
]
```

Thing is, `select2_from_ajax` is the name of the related in the relationship:

```php
    public function article()
    {
        return $this->belongsTo(\Backpack\NewsCRUD\app\Models\Article::class, 'select2_from_ajax');
    }
```
So in the end, if you are working with a relationship that should be the correct name for the relation to work. 

Maybe I did not fully understand all the use cases for this field. 
So, following the code:

```php
        $entity_model = $crud->model;

        // here it's the only place where entity is used. 
        $related_model = $crud->getRelationModel($field['entity']);
       
/*
could we assume that if entity is not set, use the defined $field['model'] ?
$related_model = !isset($field['entity']) ? $field['model'] : $crud->getRelationModel($field['entity']); 
*/

        $group_by_model = (new $related_model)->{$field['group_by']}()->getRelated();
        $categories = $group_by_model::with($field['group_by_relationship_back'])->get();
        
        // if field model is not set, we don't have categoryless.
        if (isset($field['model'])) {
            $categorylessEntries = $related_model::doesnthave($field['group_by'])->get();
        }

     ........
     //again if field model is not set nothing is run.
   @if (isset($field['model']) && isset($field['group_by']))
                @foreach ($categories as $category)
```

So, like I write in the code we could remove the entity from field if we don't want to work with a relation. That way nothing gets overriden in crud.

Let me know.

**Comment by tabacitu** [comment]
> It happens because the field name changes, but well, field name changes because it's the name that is used by laravel in the relationship.

Indeed. And _usually_ that should be fine. But sometimes people want to define their own thing. For example, you'd like to be able to select a `category` from the relationship, but instead of storing it as the `category_id`, you'd like to store it in `fallback_category_id`, or in the `extras` JSON, as `extras->secondary_category_id`.

This was possible in 4.0. And it's what I was using in MonsterCrudController. But right now, with the guessing merged, it's impossible in 4.1.

**Backpack should not stand in the way, if you're trying to do something**. It should not overwrite things, when you know what you're doing. It's an architectural decision that I'm adamant about. Yes it should help fill in the blanks. But only overwrite stuff if we're very sure that's what the developer intended.

> Maybe I did not fully understand all the use cases for this field.

Well in 4.0 developers can point a select2 to a relationship, but then store it somewhere else, by just specifying a different name. The same way they would do with any other field.

And they can have multiple selects on the same page, pointing to the same relationship. Because `entity` is the one defining the `relationship`, not `name`.

In 4.1 this is no longer possible, because the name gets overwritten by the guessing logic. Am I correct @pxpm ?

---

> here it's the only place where entity is used. 

It's also used in the fields themselves. In the blade files.

> could we assume that if entity is not set, use the defined $field['model']

Hmmm this is interesting. But I see it exactly _the other way around_. **If entity _is_ set, then do _not_ overwrite the name.** It means the person know what he's doing. And we can remove `entity` entirely from the `relationship` field docs, so people never define it there. So that for 1-n relations the name can be guessed. **Do you think this could work @pxpm ?**

But then, it would mean this `entity` (which is was badly-named to begin with, it represents the relationship on the model, so `entity` is a bad name for that), will now take an extra meaning, "do not overwrite name". So one attribute, `entity` will mean two things:
- this is the name of the relationship on the model;
- do not overwrite name to make my life easier;

All that inside one badly-chosen name, `entity`... that's pretty cryptic. Not sure this is a good solution... But it's maybe _a_ solution, and _a_ solution is better than _no_ solution 🥇 

----

I'll take a deeper dive into this right now - maybe I can figure out a better solution. As of this moment, **this basically breaks compatibility between 4.0 and 4.1 in a few niche use cases**:
- **select fields can no longer be fake**
- **select fields can no longer store their input somewhere else**

Cover me, I'm going in 🤣  If you could point me to the exact place where the `name` is getting overwritten in this exact instance, that would help a lot 🙄

**Comment by pxpm** [comment]
I agree with you. 

I think this starts to get messy because now, imagine you have a belongsTo with `article` you could do only: `$this->crud->addField('article')`; 

So technically we tied the name with the possible relation. There is no field `article` in crud, atmost there is a `article_id`.

What I don't understand is why we would need the 'entity' and also the 'model'. Without model nothing will work, so we need to provide the model. Would the model ever be different from the one we get from entity ? They only way I say it's possible to have different `$related_model` is that you provide in entity `address.icon` where current model has a relation with address that has a relation with icon, in this scenario: `$crud->getRelationModel('address.icon');` the model returned would be `App\Models\Icon` so you will get the related categories from Icon Model. The same will happen if you only setup model as `App\Models\Icon` and don't use the entity. 

I don't know if I am making myself clear here. I just don't see the point of using entity here if you don't provide the model with the same result expected from `whatever.entity.you.use`

I am feeling dumb or lost, let devil come and choose!

**Comment by tabacitu** [comment]
Until we figure out a better way to do this, I've commented out `makeSureFieldNameMatchesRelation()` in https://github.com/Laravel-Backpack/CRUD/pull/2768

**Comment by tabacitu** [comment]
> I think this starts to get messy because now, imagine you have a belongsTo with article you could do only: $this->crud->addField('article');

I've just tried it after https://github.com/Laravel-Backpack/CRUD/pull/2769 with:
```php
$this->crud->addField('category_id'); // shows ok; saves ok;
$this->crud->addField('category'); // shows ok; saves ok;

// The only difference between the two of them is the VALIDATION. 
// Inside the request, for one you need the validation rule to be for 
// 'category_id', for the other you need it to be 'category'
// But I think that's actually better - if you write `category` in the controller 
// you should write `category` in the request, it's more intuitive.
```

So they both work in my tests (with the caveat above). Please take a look at #2769 

> What I don't understand is why we would need the 'entity' and also the 'model'. Without model nothing will work, so we need to provide the model. Would the model ever be different from the one we get from entity ? They only way I say it's possible to have different $related_model is that you provide in entity address.icon where current model has a relation with address that has a relation with icon, in this scenario: $crud->getRelationModel('address.icon'); the model returned would be App\Models\Icon so you will get the related categories from Icon Model. The same will happen if you only setup model as App\Models\Icon and don't use the entity.
>
> I don't know if I am making myself clear here. I just don't see the point of using entity here if you don't provide the model with the same result expected from whatever.entity.you.use

Ugh. I didn't get this sorry 😄 Please rephrase?

**Comment by tabacitu** [comment]
This was ultimately fixed by #2769 . So let's close it.
